### PR TITLE
Item 7295: Publish and version API artifacts independently

### PR DESCRIPTION
**Pull request by labkey-susanh**
#### Rationale
Since our APIs change much less frequently than the LabKey Server versions, we should be publishing them independently of our server releases.  Publishing them to Artifactory also means the are available for reference and download there without a separate distribution needed to collect all these artifacts.  We need changes to the pushToDownload script before we can remove that distribution entirely, though.  This is a first step.

#### Related Pull Requests
* https://github.com/LabKey/labkey-api-jdbc/pull/9

#### Changes
* Added CHANGELOG and README documentation 
* renamed directories to correspond to the published artifact names as gradle expects (so TeamCity publishing task works better)
* Add publishing task and artifacts for sas api

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# LabKey Server Java and SAS API
+
+This repository contains the source for the java and SAS API artifacts for LabKey Server.
```
Comment (labkey-nicka) [inline README.md]:
```suggestion
This repository contains the source for the Java and SAS API artifacts for LabKey Server.
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# LabKey Server Java and SAS API
+
+This repository contains the source for the java and SAS API artifacts for LabKey Server.  
+
+If doing development in this repository in conjunction with the overall LabKey Server Gradle project,
```
Comment (labkey-nicka) [inline README.md]:
```suggestion
If you're doing development in this repository in conjunction with the overall LabKey Server Gradle project,
```

File: labkey-client-api/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# The LabKey Remote API Library for Java
+
+Welcome to the LabKey Remote API library for Java. This library allows Java 
+developers to access the data and services exposed from a LabKey Server installation.
+
+For more information on the library, 
+see our [JavaAPI](https://www.labkey.org/Documentation/wiki-page.view?name=javaAPI) page,
```
Comment (labkey-nicka) [inline labkey-client-api/README.md]:
```suggestion
see our [Java API](https://www.labkey.org/Documentation/wiki-page.view?name=javaAPI) page, 
```

File: labkey-client-api/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+### Dependency Declaration
+To declare a dependency on this jar file, you can use the following in Gradle
+
+```compile(group: 'org.labkey.api', name: 'labkey-client-api', version: '1.2.0')```
+
+If using the LabKey Gradle plugins and building a LabKey module, it is best to 
+use this utility method instead to facilitate testing of any local changes to
+this jar file with that module
+```
```
Comment (labkey-nicka) [inline labkey-client-api/README.md]:
```suggestion
```gradle
```

File: labkey-client-api/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+- Clone this repository (`labkey-api-java`) into the `$LABKEY_ROOT/remoteapi` directory
+- Update the version in the appropriate `build.gradle` file to a unique SNAPSHOT version. Make sure 
+it is a version number not in use by another branch, including `develop`.  A good practice to 
+follow is to use your feature branch in the naming of that version (e.g., `X.Y.Z-myNewFeature-SNAPSHOT`). 
+TeamCity will automatically publish new SNAPSHOT versions from feature branches.
+- Update your `$LABKEY_ROOT/settings.gradle` file to include this project (`include ':remoteapi:labkey-api-java:labkey-client-api'`)
+
+After making these changes, in all places where a dependency on the Java API has been declared like so: 
+
+`BuildUtils.addLabKeyDependency(project: project, config: "remoteApi", depProjectPath: BuildUtils.getRemoteApiProjectPath(gradle), depVersion: project.labkeyClientApiVersion)`
```
Comment (labkey-nicka) [inline labkey-client-api/README.md]:
```suggestion
```gradle
BuildUtils.addLabKeyDependency(project: project, config: "remoteApi", depProjectPath: BuildUtils.getRemoteApiProjectPath(gradle), depVersion: project.labkeyClientApiVersion)
```
```

File: labkey-client-api/README.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+- Update the [change log](CHANGELOG.md) to document what has changed.  You can leave the release date and version alone at this 
+point; they will get updated when the next release is published.
+- **TBD** Run tests using your SNAPSHOT version of the tests
+- Merge your branch into develop if appropriate tests are passing.
+
+
+### Publishing
+
+Information about the process for publishing new versions of this library
+can be found in our [internal documentation](https://internal.labkey.com/Handbook/Dev/wiki-page.view?name=mavenArtifacts)
```
Comment (labkey-nicka) [inline labkey-client-api/README.md]:
```suggestion
can be found in our [internal documentation](https://internal.labkey.com/Handbook/Dev/wiki-page.view?name=mavenArtifacts).
```

**Review by labkey-nicka — CHANGES_REQUESTED** [review]
I haven't done a full pass of following the documented directions but so far this looks great. I'll circle back with any additional feedback after I go through the steps.

File: labkey-client-api/README.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+### Making Changes
+
+Though this repository contains its own Gradle version and settings file and can be built
+independently of the rest of LabKey server, you can also include it in the larger LabKey 
+Server Gradle project if you need to make changes to the code and test them locally without
+having to publish.  
+
+- Create a feature branch of this repository (`fb_myNewFeature`)
```
Comment (labkey-nicka) [inline labkey-client-api/README.md]: nit: These first two steps seem out of order to me. 1) Clone the repo, 2) Make a feature branch.

File: labkey-client-api/README.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
+# The LabKey Remote API Library for Java
+
+Welcome to the LabKey Remote API library for Java. This library allows Java
```
Comment (labkey-nicka) [inline labkey-client-api/README.md]: nit: As welcomed as I feel I don't think this verbiage is necessary.

```suggestion
This library allows Java 
```

**Review by labkey-nicka — APPROVED** [review]
Steps work well. My feedback is all nits so approved! I don't know anything about SAS except that they apparently like their scriptstobealllowercase.sas. I'll leave that to @labkey-adam.

File: labkey-client-api/README.md (hunk 7 above)
Comment (labkey-susanh) [reply labkey-client-api/README.md]: Depends on how you usually make your feature branches, I guess, but sure.

File: labkey-client-api/README.md (hunk 8 above)
Comment (labkey-susanh) [reply labkey-client-api/README.md]: Yeah, it seemed odd to me too.  Gone now.

File: labkey-api-sas/README.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# The LabKey Remote API Library for SAS
+
+Library of macros for use in accessing LabKey Server via SAS.
+
+For more information on the library, including setup instructions and
+examples, see our [SAS API](https://www.labkey.org/Documentation/wiki-page.view?name=sasAPI)
+documentation.
+
+If you have any questions about this library or the LabKey Server, 
+see our [Community Support Forums](https://www.labkey.org/home/Support/LabKey%20Support%20Forum/project-begin.view?).
```
Comment (labkey-adam) [inline labkey-api-sas/README.md]: I'd remove these two lines. We don't really want to support this library on the community forums

File: labkey-client-api/README.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+This library allows Java 
+developers to access the data and services exposed from a LabKey Server installation.
+
+For more information on the library, 
+see our [Java API](https://www.labkey.org/Documentation/wiki-page.view?name=javaAPI) page, 
+which includes a link to the [JavaDoc](https://www.labkey.org/download/clientapi_docs/java-api/)
+where you can get details about the classes and programming interfaces. 
+
+If you have any questions about this library or the LabKey Server,
```
Comment (labkey-adam) [inline labkey-client-api/README.md]:
```suggestion
If you have any questions about this library or LabKey Server, 
```

File: labkey-client-api/build.gradle [hunk 11]
```diff
@@ -54,10 +55,9 @@ repositories {
     }
 }
 
+// TODO remove this and use defualt once client-api distribution doesn't need it
```
Comment (labkey-adam) [inline labkey-client-api/build.gradle]:
```suggestion
// TODO remove this and use default once client-api distribution doesn't need it
```

**Review by labkey-adam — APPROVED** [review]
Minor nits in comments - not required

**Pull request merged**